File: leetcode/leetcode_86_partition_list/Solution.java
package leetcode_86_partition_list;


//Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}

class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode p = head;
        ListNode dummy_small = new ListNode();
        ListNode dummy_big = new ListNode();
        ListNode p1 = dummy_small;
        ListNode p2 = dummy_big;

        while (p != null){
            if (p.val < x){
                p1.next = p;
                p1 = p1.next;
            }else{
                p2.next = p;
                p2 = p2.next;
            }

            ListNode temp = p.next;
            p.next = null;
            p = temp;
        }

        p1.next = dummy_big.next;
        return dummy_small.next;
        
    }
}

File: leetcode/leetcode_86_partition_list/solution.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def partition(self, head: Optional[ListNode], x: int) -> Optional[ListNode]:
        p = head
        dummy_small = ListNode()
        dummy_big = ListNode()
        p1 = dummy_small
        p2 = dummy_big

        while p != None:
            if p.val >= x:
                p2.next = p
                p2 = p2.next
            else:
                p1.next = p
                p1 = p1.next

            temp = p.next
            p.next = None
            p = temp

        p1.next = dummy_big.next
        return dummy_small.next

